File: 100/198 - House Robber.c
#define max(a, b) ((a) > (b) ? (a) : (b))

int rob(int* nums, int numsSize) {
    int *dp, i, ret;
    
    if (! numsSize)
        return 0;
    if (numsSize == 1)
        return nums[0];
    dp = (int *)malloc(numsSize * sizeof(int));
    dp[0] = nums[0];
    dp[1] = max(nums[0], nums[1]);
    
    for (i = 2; i < numsSize; ++i)
        dp[i] = max(nums[i] + dp[i - 2], dp[i - 1]);
    
    ret = dp[numsSize - 1];
    free(dp);
    
    return ret;
}

File: csharp/198.house-robber.cs
/*
 * @lc app=leetcode id=198 lang=csharp
 *
 * [198] House Robber
 */
public class Solution {
    public int Rob(int[] nums) {
        int prev = 0, ret = 0, curr = 0;
        foreach (int num in nums) {
            int rob = prev + num;
            ret = Math.Max(ret, rob);
            prev = Math.Max(prev, curr);
            curr = rob;
        }
        return ret;
    }
}

